### ISSUE TITLE
wallpaper will not load, `terminate called after throwing an instance of 'std::runtime_error'`

### ISSUE THREAD

**Issue by Laar3** [Almamu/linux-wallpaperengine#59]
Hi, when i try to load a wallpaper i receive the following error
![image](https://user-images.githubusercontent.com/77225642/143675382-b945fed9-8767-448e-8019-8bfabea3ff0b.png)
i also received this warning during compilation but cant see any dependencies i'm missing. 
![image](https://user-images.githubusercontent.com/77225642/143675468-b6218a2a-e92d-452e-a972-8db0555bb015.png)

**Comment by Almamu** [comment]
Hello @Laar3
About the cmake error, you can actually ignore it, it's due to how the variables for CMAKE are named, nothing important.

I've had no issue loading the background you're trying to load. Please make sure you're running the latest version from the "main" branch. If you still have issues, can you please take a backtrace of the program? (with something like gdb would be enough) so I can know what exactly is going on?

**Comment by Laar3** [comment]
yep i'm running the latest version from the main branch, this is the output from gdb.
![image](https://user-images.githubusercontent.com/77225642/143944666-2d047405-dae8-48b7-9d31-f516edf7cdbc.png)
when running from gdb it draws a window (top right in the screenshot)
![image](https://user-images.githubusercontent.com/77225642/143944795-a184d4d1-b115-4822-8c25-b383727af117.png)
however it takes on the look of whatever was in that spot previously(Now its the terminal on the left 
![image](https://user-images.githubusercontent.com/77225642/143944908-7df20379-ebd2-4864-9815-7068aba25f33.png)
)

**Comment by Almamu** [comment]
Hi @Laar3 
If you're still getting that error (Expected type) with that background It means your repo is out of date. I can run that exact same background on the last commit uploaded. Please run
```
git pull
```

and re-run the cmake steps again and your background should load just fine. This commit (9d2e9b43ee2eeb772ee65c0c9a8d72018ad62887) fixed some issues with shader variables, and that one fixed the background you're trying to load.

**Comment by Laar3** [comment]
ah, sorry i thought i had the latest version. one issue though, the wallpaper normally has sparks and fog however they are not displaying. I've copied the assets folder in so it is in the same directory as the wallengine executable.

**Comment by Almamu** [comment]
Hi @Laar3 
It'll depend on what the background is using for those effects. If they're particles (which sounds like the most likely scenario), support for them is not implemented yet, so you'll only see normal effects like water, waving flags, etc

**Comment by hackaholic5069** [comment]
im having a similar issue, and i only cloned and compiled wallengine today
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Cannot find the file in any of the containers
[1]    11231 IOT instruction (core dumped)  ./wallengine ~/Documents/wallpapers/hk
```

**Comment by SeanOMik** [comment]
> im having a similar issue, and i only cloned and compiled wallengine today
> 
> ```
> terminate called after throwing an instance of 'std::runtime_error'
>   what():  Cannot find the file in any of the containers
> [1]    11231 IOT instruction (core dumped)  ./wallengine ~/Documents/wallpapers/hk
> ```

The same thing happened to me also. I'm running Arch

**Comment by Almamu** [comment]
@hackaholic5069 @SeanOMik For that I'll need to know what background you're running as these kind of errors can be a bit hard to debug. Also, please make sure the default assets folder is properly setup too.

**Comment by hackaholic5069** [comment]
it happened with every wallpaper ive tried
i just went onto the popular page and downloaded a few random ones, and all of them failed with the same error

**Comment by SeanOMik** [comment]
[This](https://steamcommunity.com/sharedfiles/filedetails/?id=2034969129&searchtext=Future+On+mars) is the wallpaper I used. I didn't think about trying other ones at the time.

I think I'll have to reinstall this, but I'll be willing to try with a wallpaper that is known to work.

**Comment by iamcodernitesh** [comment]
@Almamu Same issue when used multiple wallpapers. Even tried the one in the README.md file but no luck.

**Comment by Almamu** [comment]
@SeanOMik that wallpaper you've linked seems to work just fine on my end. Could you provide some extra details on your setup? (distro and gpu should suffice so I can set up a testbed with those.

@iamcodernitesh that's really weird, I haven't had any issues with the wallpaper mentioned in the README.md.

I'm thinking this could be an issue related either to the library versions installed on your systems or specific to the distro you're using. @iamcodernitesh would you be able to run the program with gdb (or an IDE w/ debugger) and get a stacktrace of where the crash happens? The ```bt``` command in the gdb commandline should give a backtrace after the app crashes.

**Comment by SeanOMik** [comment]
> @SeanOMik that wallpaper you've linked seems to work just fine on my end. Could you provide some extra details on your setup? (distro and gpu should suffice so I can set up a testbed with those).

I'm using Arch, an Nvidia GeForce 970, with an AMD Ryzen 5 3600 CPU. 


> I'm thinking this could be an issue related either to the library versions installed on your systems or specific to the distro you're using.

It could very well be a library version issue since I'm using Arch.

**Comment by Almamu** [comment]
I've added some extra logging information on commit 6aab83a6972fc43bd41496e83a07d280555bd6da so I can get a better picture of how the app is being run and what is being detected.

**Comment by Almamu** [comment]
> It could very well be a library version issue since I'm using Arch.

My main development machine is Arch too, so unless you've done some manual installation of libraries, it should just work :/ I'll need some extra information from the last build to take a closer look at what's happening, just to make sure that I'm properly understanding the issue.

**Comment by SeanOMik** [comment]
> I've added some extra logging information on commit [6aab83a](https://github.com/Almamu/linux-wallpaperengine/commit/6aab83a6972fc43bd41496e83a07d280555bd6da) so I can get a better picture of how the app is being run and what is being detected.

For some reason with that commit, it works fine. Maybe there were some other commits that were pushed since I last built that fixed it, or maybe it was some weird library version issue, and upgrading the system since then fixed it (upgraded it about a week ago for something else).

I do get an error when trying to use `--screen-root`, but I can post it in a separate issue if you'd like.
```
Found requested screen: DVI-D-0 -> 1920x0:1920x1080
X Error of failed request:  BadMatch (invalid parameter attributes)
  Major opcode of failed request:  152 (GLX)
  Minor opcode of failed request:  5 (X_GLXMakeCurrent)
  Serial number of failed request:  151
  Current serial number in output stream:  151
```

**Comment by C0nvert** [comment]
Have the same Error as @SeanOMik 
What Debug Log do you need?
journalctl do not have anything interesting in there.

**Comment by C0nvert** [comment]
@Almamu 
Sorry if you got it wrong.
The missing scene.pkg quirk was never a problem for me on the newest commit.
I only tried it, cause it worked for @SeanOMik above.





> For some reason with that commit, it works fine. Maybe there were some other commits that were pushed since I last built that fixed it, or maybe it was some weird library version issue, and upgrading the system since then fixed it (upgraded it about a week ago for something else).


I will run the Binary on it's own and report back to you.
On an important note, i'm too using Nvidia's proprietary drivers.
Maybe it's kinda connected.

**Comment by Almamu** [comment]
@C0nvert 
> On an important note, i'm too using Nvidia's proprietary drivers.
> Maybe it's kinda connected.

If you're getting an error like this when running the binary with the --screen-root switch, then yeah, most likely that's the cause:
```
Found requested screen: DVI-D-0 -> 1920x0:1920x1080
X Error of failed request:  BadMatch (invalid parameter attributes)
  Major opcode of failed request:  152 (GLX)
  Minor opcode of failed request:  5 (X_GLXMakeCurrent)
  Serial number of failed request:  151
  Current serial number in output stream:  151
```

> The missing scene.pkg quirk was never a problem for me on the newest commit, only missing json.project file.

For that, could you do an ```ls -la``` of the folder where the scene.pkg file is? There has to be a project.json alongside it, otherwise the app won't know what type of background it is. If It is, but cannot be found can you check the file permissions and make sure it can be read by the user that is running the ```wallengine``` binary?

**Comment by C0nvert** [comment]
> If you're getting an error like this when running the binary with the --screen-root switch, then yeah, most likely that's the cause:

Just getting same Error as before:
```
Detected scene.pkg file at /home/alex/Downloads/2596694556/scene.pkg. Adding to list of searchable paths
terminate called after throwing an instance of 'WallpaperEngine::Assets::CAssetLoadException'
  what():  Cannot find file project.json: Cannot find file in any of the containers
zsh: abort (core dumped)  ./wallengine --screen-root DVI-D-0 /home/alex/Downloads/2596694556
```


> For that, could you do an `ls -la` of the folder where the scene.pkg file is? There has to be a project.json alongside it, otherwise the app won't know what type of background it is. If It is, but cannot be found can you check the file permissions and make sure it can be read by the user that is running the `wallengine` binary?

Seems okay for me. Running as User:
```
drwxr-xr-x 2 alex alex    4096 18. Feb 14:14 .
drwxr-xr-x 5 alex alex    4096 18. Feb 16:30 ..
-rwxr-xr-x 1 alex alex  560218 18. Feb 14:05 preview.gif
-rwxr-xr-x 1 alex alex     613 18. Feb 14:05 project.json
-rwxr-xr-x 1 alex alex 8193125 18. Feb 14:05 scene.pkg

```

I also changed it to 777 just for testing purposes:
```
drwxr-xr-x 2 alex alex     4096 18. Feb 14:14 .
drwxr-xr-x 6 alex alex     4096 18. Feb 16:41 ..
-rwxrwxrwx 1 alex alex   620658 18. Feb 14:01 preview.jpg
-rwxrwxrwx 1 alex alex     1550 18. Feb 14:01 project.json
-rwxrwxrwx 1 alex alex 40353370 18. Feb 14:01 scene.pkg

```

Same Result:
```
Detected scene.pkg file at 2596694556/scene.pkg. Adding to list of searchable paths
terminate called after throwing an instance of 'WallpaperEngine::Assets::CAssetLoadException'
  what():  Cannot find file project.json: Cannot find file in any of the containers
zsh: abort (core dumped)  ./wallengine 2596694556
```

**Comment by Almamu** [comment]
Okay, now I see why It's not working right for you. Looks like I missed some things on the path fixes so trailing slashes are not properly handled. I've uploaded some changes (d5745b1cfaab2dc51374d7fbd05ddf932ad90348) that should fix this issue for you.

I'm not really sure it'll work fine on GNOME. Sadly I haven't had much time to implement support for DE, so the only easy and tested way is to run without compositor/something drawing to the X background.

**Comment by C0nvert** [comment]
@Almamu can confirm this fixes it.
Thank you for your hard work!!!! :+1: 
But now i have this kind of artifacts, they differ wich each Wallpaper.
![Wallpaper](https://user-images.githubusercontent.com/16489340/154772881-0f49e734-2c8e-4dd6-a244-3597f36d5124.png)

Any suggestions?

**Comment by Almamu** [comment]
@C0nvert I've taken a look at the issue and identified the problem. I've added a quick fix for this on the latest commit (131801b7f09e50e33ad64cc0f6782ca4e5665d8b) that should fix it. Although I don't know if It'll fix all the backgrounds that have a similar issue. Please open a new issue with any of the artifacts you find and I'll take a look into it.
